[Improvement] in HTTPClient.java character set should be specified

### What would you like to be improved?

HTTPClient.java:336: warning: [DefaultCharset] Implicit use of the platform default charset, which can result in differing behaviour between JVM executions or incorrect behavior if the encoding of the data source doesn't match expectations.
```java
AuthConstants.HTTP_HEADER_AUTHORIZATION, new String(authDataProvider.getTokenData()));
```

 Did you mean 'AuthConstants.HTTP_HEADER_AUTHORIZATION, new String(authDataProvider.getTokenData(), UTF_8));' or 'AuthConstants.HTTP_HEADER_AUTHORIZATION, new String(authDataProvider.getTokenData(), Charset.defaultCharset()));'?

See https://errorprone.info/bugpattern/DefaultCharset

### How should we improve?

Select one to use

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -296,7 +296,8 @@ subprojects {
         "UnusedCollectionModifiedInPlace",
         "VarTypeName",
         "XorPower",
-        "EqualsGetClass"
+        "EqualsGetClass",
+        "DefaultCharset"
       )
     }
   }

--- a/catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/miniHMS/MiniHiveMetastore.java
+++ b/catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/miniHMS/MiniHiveMetastore.java
@@ -30,6 +30,7 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.io.Reader;
+import java.nio.charset.StandardCharsets;
 import java.sql.Connection;
 import java.sql.DriverManager;
 import java.sql.SQLException;
@@ -142,7 +143,7 @@ public class MiniHiveMetastore {
 
     ClassLoader classLoader = ClassLoader.getSystemClassLoader();
     InputStream inputStream = classLoader.getResourceAsStream("hive-schema-3.1.0.derby.sql");
-    try (Reader reader = new InputStreamReader(inputStream)) {
+    try (Reader reader = new InputStreamReader(inputStream, StandardCharsets.UTF_8)) {
       scriptRunner.runScript(reader);
     }
   }

--- a/catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/miniHMS/ScriptRunner.java
+++ b/catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/miniHMS/ScriptRunner.java
@@ -18,10 +18,13 @@
  */
 package com.datastrato.gravitino.catalog.hive.miniHMS;
 
+import java.io.BufferedWriter;
 import java.io.IOException;
 import java.io.LineNumberReader;
+import java.io.OutputStreamWriter;
 import java.io.PrintWriter;
 import java.io.Reader;
+import java.nio.charset.StandardCharsets;
 import java.sql.Connection;
 import java.sql.ResultSet;
 import java.sql.ResultSetMetaData;
@@ -40,11 +43,13 @@ public class ScriptRunner {
   private final boolean stopOnError;
   private final boolean autoCommit;
 
-  @SuppressWarnings("DefaultCharset")
-  private final PrintWriter logWriter = new PrintWriter(System.out);
+  private final PrintWriter logWriter =
+      new PrintWriter(
+          new BufferedWriter(new OutputStreamWriter(System.out, StandardCharsets.UTF_8)));
 
-  @SuppressWarnings("DefaultCharset")
-  private final PrintWriter errorLogWriter = new PrintWriter(System.err);
+  private final PrintWriter errorLogWriter =
+      new PrintWriter(
+          new BufferedWriter(new OutputStreamWriter(System.err, StandardCharsets.UTF_8)));
 
   private String delimiter = DEFAULT_DELIMITER;
   private boolean fullLineDelimiter = false;

--- a/catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/miniHMS/ScriptRunner.java
+++ b/catalogs/catalog-hive/src/test/java/com/datastrato/gravitino/catalog/hive/miniHMS/ScriptRunner.java
@@ -40,7 +40,10 @@ public class ScriptRunner {
   private final boolean stopOnError;
   private final boolean autoCommit;
 
+  @SuppressWarnings("DefaultCharset")
   private final PrintWriter logWriter = new PrintWriter(System.out);
+
+  @SuppressWarnings("DefaultCharset")
   private final PrintWriter errorLogWriter = new PrintWriter(System.err);
 
   private String delimiter = DEFAULT_DELIMITER;

--- a/clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java
+++ b/clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java
@@ -34,6 +34,7 @@ import java.io.IOException;
 import java.io.UncheckedIOException;
 import java.net.URI;
 import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
 import java.util.Map;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
@@ -333,7 +334,8 @@ public class HTTPClient implements RESTClient {
     }
     if (authDataProvider != null) {
       request.setHeader(
-          AuthConstants.HTTP_HEADER_AUTHORIZATION, new String(authDataProvider.getTokenData()));
+          AuthConstants.HTTP_HEADER_AUTHORIZATION,
+          new String(authDataProvider.getTokenData(), StandardCharsets.UTF_8));
     }
 
     try (CloseableHttpResponse response = httpClient.execute(request)) {

--- a/clients/client-java/src/main/java/com/datastrato/gravitino/client/SimpleTokenProvider.java
+++ b/clients/client-java/src/main/java/com/datastrato/gravitino/client/SimpleTokenProvider.java
@@ -27,7 +27,8 @@ final class SimpleTokenProvider implements AuthDataProvider {
     this.token =
         (AuthConstants.AUTHORIZATION_BASIC_HEADER
                 + new String(
-                    Base64.getEncoder().encode(userInformation.getBytes(StandardCharsets.UTF_8))))
+                    Base64.getEncoder().encode(userInformation.getBytes(StandardCharsets.UTF_8)),
+                    StandardCharsets.UTF_8))
             .getBytes(StandardCharsets.UTF_8);
   }
 

--- a/clients/client-java/src/test/java/com/datastrato/gravitino/client/TestOAuth2TokenProvider.java
+++ b/clients/client-java/src/test/java/com/datastrato/gravitino/client/TestOAuth2TokenProvider.java
@@ -19,6 +19,7 @@ import io.jsonwebtoken.Jwts;
 import io.jsonwebtoken.SignatureAlgorithm;
 import io.jsonwebtoken.security.Keys;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.security.KeyPair;
 import java.util.Date;
 import org.apache.hc.core5.http.HttpStatus;
@@ -134,6 +135,6 @@ public class TestOAuth2TokenProvider {
     Assertions.assertNotEquals(accessToken, oldAccessToken);
     Assertions.assertEquals(
         AuthConstants.AUTHORIZATION_BEARER_HEADER + accessToken,
-        new String(provider.getTokenData()));
+        new String(provider.getTokenData(), StandardCharsets.UTF_8));
   }
 }

--- a/clients/client-java/src/test/java/com/datastrato/gravitino/client/TestSimpleTokenProvider.java
+++ b/clients/client-java/src/test/java/com/datastrato/gravitino/client/TestSimpleTokenProvider.java
@@ -18,7 +18,7 @@ public class TestSimpleTokenProvider {
     try (AuthDataProvider provider = new SimpleTokenProvider()) {
       Assertions.assertTrue(provider.hasTokenData());
       String user = System.getenv("GRAVITINO_USER");
-      String token = new String(provider.getTokenData());
+      String token = new String(provider.getTokenData(), StandardCharsets.UTF_8);
       Assertions.assertTrue(token.startsWith(AuthConstants.AUTHORIZATION_BASIC_HEADER));
       String tokenString =
           new String(
@@ -26,7 +26,8 @@ public class TestSimpleTokenProvider {
                   .decode(
                       token
                           .substring(AuthConstants.AUTHORIZATION_BASIC_HEADER.length())
-                          .getBytes(StandardCharsets.UTF_8)));
+                          .getBytes(StandardCharsets.UTF_8)),
+              StandardCharsets.UTF_8);
       if (user != null) {
         Assertions.assertEquals(user + ":dummy", tokenString);
       } else {

--- a/core/src/main/java/com/datastrato/gravitino/storage/kv/KvEntityStore.java
+++ b/core/src/main/java/com/datastrato/gravitino/storage/kv/KvEntityStore.java
@@ -439,7 +439,7 @@ public class KvEntityStore implements EntityStore {
         return StorageLayoutVersion.V1;
       }
 
-      return StorageLayoutVersion.fromString(new String(bytes));
+      return StorageLayoutVersion.fromString(new String(bytes, StandardCharsets.UTF_8));
     } catch (IOException e) {
       throw new IllegalStateException("Failed to get/put layout version information", e);
     }

--- a/core/src/main/java/com/datastrato/gravitino/storage/kv/KvNameMappingService.java
+++ b/core/src/main/java/com/datastrato/gravitino/storage/kv/KvNameMappingService.java
@@ -68,7 +68,7 @@ public class KvNameMappingService implements NameMappingService {
     return FunctionUtils.executeInTransaction(
         () -> {
           byte[] name = transactionalKvBackend.get(idByte);
-          return name == null ? null : new String(name);
+          return name == null ? null : new String(name, StandardCharsets.UTF_8);
         },
         transactionalKvBackend);
   }

--- a/core/src/test/java/com/datastrato/gravitino/storage/kv/TestKvGarbageCollector.java
+++ b/core/src/test/java/com/datastrato/gravitino/storage/kv/TestKvGarbageCollector.java
@@ -34,6 +34,7 @@ import com.datastrato.gravitino.storage.TransactionIdGenerator;
 import com.datastrato.gravitino.storage.kv.KvGarbageCollector.LogHelper;
 import java.io.File;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.time.Instant;
 import java.util.List;
@@ -42,6 +43,7 @@ import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 import org.mockito.Mockito;
 
+@SuppressWarnings("DefaultCharset")
 class TestKvGarbageCollector {
   public Config getConfig() throws IOException {
     Config config = Mockito.mock(Config.class);
@@ -192,8 +194,8 @@ class TestKvGarbageCollector {
       List<Pair<byte[], byte[]>> data =
           kvBackend.scan(
               new KvRange.KvRangeBuilder()
-                  .start("_".getBytes())
-                  .end("z".getBytes())
+                  .start("_".getBytes(StandardCharsets.UTF_8))
+                  .end("z".getBytes(StandardCharsets.UTF_8))
                   .startInclusive(false)
                   .endInclusive(false)
                   .build());

--- a/core/src/test/java/com/datastrato/gravitino/storage/kv/TestTransactionalKvBackend.java
+++ b/core/src/test/java/com/datastrato/gravitino/storage/kv/TestTransactionalKvBackend.java
@@ -39,6 +39,7 @@ import org.mockito.Mockito;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+@SuppressWarnings("DefaultCharset")
 class TestTransactionalKvBackend {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(TestTransactionalKvBackend.class);

--- a/integration-test/src/main/java/com/datastrato/gravitino/integration/test/util/ITUtils.java
+++ b/integration-test/src/main/java/com/datastrato/gravitino/integration/test/util/ITUtils.java
@@ -8,6 +8,7 @@ import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Paths;
 import java.util.Map;
@@ -36,7 +37,7 @@ public class ITUtils {
       for (String key : props.stringPropertyNames()) {
         String value = props.getProperty(key);
         // Use customized write functions to avoid escaping `:` into `\:`.
-        outputStream.write((key + " = " + value + "\n").getBytes());
+        outputStream.write((key + " = " + value + "\n").getBytes(StandardCharsets.UTF_8));
       }
     }
   }

--- a/integration-test/src/main/java/com/datastrato/gravitino/integration/test/util/OAuthMockDataProvider.java
+++ b/integration-test/src/main/java/com/datastrato/gravitino/integration/test/util/OAuthMockDataProvider.java
@@ -6,6 +6,7 @@ package com.datastrato.gravitino.integration.test.util;
 
 import com.datastrato.gravitino.client.OAuth2TokenProvider;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 
 public class OAuthMockDataProvider extends OAuth2TokenProvider {
 
@@ -15,7 +16,7 @@ public class OAuthMockDataProvider extends OAuth2TokenProvider {
 
   @Override
   protected String getAccessToken() {
-    return new String(token);
+    return new String(token, StandardCharsets.UTF_8);
   }
 
   public static OAuthMockDataProvider getInstance() {

--- a/integration-test/src/test/java/com/datastrato/gravitino/integration/test/trino/TrinoQueryIT.java
+++ b/integration-test/src/test/java/com/datastrato/gravitino/integration/test/trino/TrinoQueryIT.java
@@ -8,6 +8,7 @@ import com.datastrato.gravitino.Namespace;
 import com.datastrato.gravitino.integration.test.util.ITUtils;
 import java.io.FileOutputStream;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashMap;
@@ -421,7 +422,7 @@ public class TrinoQueryIT extends TrinoQueryITBase {
     boolean firstLine = true;
     while (sqlMatcher.find()) {
       if (!firstLine) {
-        outputStream.write("\n".getBytes());
+        outputStream.write("\n".getBytes(StandardCharsets.UTF_8));
       }
       firstLine = false;
 
@@ -437,8 +438,8 @@ public class TrinoQueryIT extends TrinoQueryITBase {
                 + "\nresult:\n"
                 + result);
       }
-      outputStream.write(result.getBytes());
-      outputStream.write("\n".getBytes());
+      outputStream.write(result.getBytes(StandardCharsets.UTF_8));
+      outputStream.write("\n".getBytes(StandardCharsets.UTF_8));
     }
 
     outputStream.close();

--- a/integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/ProcessData.java
+++ b/integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/ProcessData.java
@@ -22,6 +22,7 @@ import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.io.PrintWriter;
 import java.io.StringWriter;
+import java.nio.charset.StandardCharsets;
 import java.util.concurrent.TimeUnit;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -178,8 +179,10 @@ public class ProcessData {
     try {
       InputStream in = this.checkedProcess.getInputStream();
       InputStream inErrors = this.checkedProcess.getErrorStream();
-      BufferedReader inReader = new BufferedReader(new InputStreamReader(in));
-      BufferedReader inReaderErrors = new BufferedReader(new InputStreamReader(inErrors));
+      BufferedReader inReader =
+          new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
+      BufferedReader inReaderErrors =
+          new BufferedReader(new InputStreamReader(inErrors, StandardCharsets.UTF_8));
       if (LOG.isTraceEnabled()) {
         LOG.trace(
             "Started retrieving data from streams of attached process: {}", this.checkedProcess);

--- a/integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/rest/OAuth2OperationsIT.java
+++ b/integration-test/src/test/java/com/datastrato/gravitino/integration/test/web/rest/OAuth2OperationsIT.java
@@ -28,7 +28,8 @@ public class OAuth2OperationsIT extends AbstractIT {
 
   private static final KeyPair keyPair = Keys.keyPairFor(SignatureAlgorithm.RS256);
   private static final String publicKey =
-      new String(Base64.getEncoder().encode(keyPair.getPublic().getEncoded()));
+      new String(
+          Base64.getEncoder().encode(keyPair.getPublic().getEncoded()), StandardCharsets.UTF_8);
 
   private static String token;
 

--- a/server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java
+++ b/server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java
@@ -18,6 +18,7 @@ import com.datastrato.gravitino.auth.KerberosUtils;
 import com.datastrato.gravitino.exceptions.UnauthorizedException;
 import com.google.common.base.Splitter;
 import java.io.File;
+import java.nio.charset.StandardCharsets;
 import java.security.Principal;
 import java.security.PrivilegedActionException;
 import java.security.PrivilegedExceptionAction;
@@ -99,7 +100,7 @@ public class KerberosAuthenticator implements Authenticator {
       throw new UnauthorizedException("Empty token authorization header", AuthConstants.NEGOTIATE);
     }
 
-    String authData = new String(tokenData);
+    String authData = new String(tokenData, StandardCharsets.UTF_8);
     if (StringUtils.isBlank(authData)
         || !authData.startsWith(AuthConstants.AUTHORIZATION_NEGOTIATE_HEADER)) {
       throw new UnauthorizedException(

--- a/server-common/src/main/java/com/datastrato/gravitino/server/auth/OAuth2TokenAuthenticator.java
+++ b/server-common/src/main/java/com/datastrato/gravitino/server/auth/OAuth2TokenAuthenticator.java
@@ -20,6 +20,7 @@ import io.jsonwebtoken.SignatureAlgorithm;
 import io.jsonwebtoken.UnsupportedJwtException;
 import io.jsonwebtoken.security.Keys;
 import io.jsonwebtoken.security.SignatureException;
+import java.nio.charset.StandardCharsets;
 import java.security.Key;
 import java.security.KeyFactory;
 import java.security.Principal;
@@ -48,7 +49,7 @@ class OAuth2TokenAuthenticator implements Authenticator {
     if (tokenData == null) {
       throw new UnauthorizedException("Empty token authorization header");
     }
-    String authData = new String(tokenData);
+    String authData = new String(tokenData, StandardCharsets.UTF_8);
     if (StringUtils.isBlank(authData)
         || !authData.startsWith(AuthConstants.AUTHORIZATION_BEARER_HEADER)) {
       throw new UnauthorizedException("Invalid token authorization header");

--- a/server-common/src/main/java/com/datastrato/gravitino/server/auth/SimpleAuthenticator.java
+++ b/server-common/src/main/java/com/datastrato/gravitino/server/auth/SimpleAuthenticator.java
@@ -8,6 +8,7 @@ package com.datastrato.gravitino.server.auth;
 import com.datastrato.gravitino.Config;
 import com.datastrato.gravitino.UserPrincipal;
 import com.datastrato.gravitino.auth.AuthConstants;
+import java.nio.charset.StandardCharsets;
 import java.security.Principal;
 import java.util.Base64;
 import org.apache.commons.lang3.StringUtils;
@@ -30,7 +31,7 @@ class SimpleAuthenticator implements Authenticator {
     if (tokenData == null) {
       return ANONYMOUS_PRINCIPAL;
     }
-    String authData = new String(tokenData);
+    String authData = new String(tokenData, StandardCharsets.UTF_8);
     if (StringUtils.isBlank(authData)) {
       return ANONYMOUS_PRINCIPAL;
     }
@@ -42,7 +43,8 @@ class SimpleAuthenticator implements Authenticator {
       return ANONYMOUS_PRINCIPAL;
     }
     try {
-      String[] userInformation = new String(Base64.getDecoder().decode(credential)).split(":");
+      String[] userInformation =
+          new String(Base64.getDecoder().decode(credential), StandardCharsets.UTF_8).split(":");
       if (userInformation.length != 2) {
         return ANONYMOUS_PRINCIPAL;
       }

--- a/server-common/src/test/java/com/datastrato/gravitino/server/auth/TestOAuth2TokenAuthenticator.java
+++ b/server-common/src/test/java/com/datastrato/gravitino/server/auth/TestOAuth2TokenAuthenticator.java
@@ -28,7 +28,9 @@ public class TestOAuth2TokenAuthenticator {
   public void testAuthentication() {
     OAuth2TokenAuthenticator auth2TokenAuthenticator = new OAuth2TokenAuthenticator();
     KeyPair keyPair = Keys.keyPairFor(SignatureAlgorithm.RS256);
-    String publicKey = new String(Base64.getEncoder().encode(keyPair.getPublic().getEncoded()));
+    String publicKey =
+        new String(
+            Base64.getEncoder().encode(keyPair.getPublic().getEncoded()), StandardCharsets.UTF_8);
     Config config = new Config(false) {};
     config.set(OAuthConfig.SERVICE_AUDIENCE, "service1");
     Assertions.assertThrows(

--- a/server-common/src/test/java/com/datastrato/gravitino/server/auth/TestSimpleAuthenticator.java
+++ b/server-common/src/test/java/com/datastrato/gravitino/server/auth/TestSimpleAuthenticator.java
@@ -46,7 +46,8 @@ public class TestSimpleAuthenticator {
                 (AuthConstants.AUTHORIZATION_BASIC_HEADER
                         + new String(
                             Base64.getEncoder()
-                                .encode("gravitino:gravitino".getBytes(StandardCharsets.UTF_8))))
+                                .encode("gravitino:gravitino".getBytes(StandardCharsets.UTF_8)),
+                            StandardCharsets.UTF_8))
                     .getBytes(StandardCharsets.UTF_8))
             .getName());
   }

